File: pw/Graphics/Prims2/Circle.py
# -*- coding: utf-8 -*-

from SceneNode import SceneNode

from OpenGL.GL import *
import math

class Circle(SceneNode):
    def geom(self):
        SpatialNode2.msg_render2()
        
        center = self.getVar("center",False)
        
        if not center:
            glTranslatef(0.5,0.5,0)
        
        if self.getVar("filled",True):
            glBegin(GL_TRIANGLE_FAN)
            glTexCoord2f(0.5,0.5)
            glVertex2f(0,0)
        else:
            glBegin(GL_LINE_LOOP)
        
        slices = self.getVar("slices",8)
        for i in range(slices+1):
            th = float(i)/float(slices)*math.pi*2
            x = math.cos(th)
            y = math.sin(th)
            glTexCoord2f((x+1)/2,(y+1)/2)
            glVertex2f(x,y)
        glEnd()

        if not center:
            glTranslatef(-0.5,-0.5,0)

File: fhe5/Graphics/Prims/Circle.cpp
#include "Circle.h"
#include <fhe/math/fheMath.h>
#include <fhe/math/Vec2.h>
#include <SDL/SDL_opengl.h>

namespace fhe
{
    namespace Graphics
    {
        FHE_ASPECT(Circle,SceneNode2);
        
        FHE_FUNC_IMPL(Circle,msg_render2)
        {
            int slices = getEntity()->getVar<int>("slices",32);
            
            if ( getEntity()->getVar<bool>("filled",true) )
            {
                glBegin(GL_TRIANGLE_FAN);
                glTexCoord2f(0.5,0.5);
                glVertex2f(0,0);
            }
            else
            {
                glBegin(GL_LINE_LOOP);
            }
            
            float th, x, y;
            
            for ( int i = 0; i < slices+1; ++i )
            {
                th = Math::TWO_PI * float(i) / float(slices);
                x = Math::cos(th);
                y = Math::sin(th);
                glTexCoord2f(0.5 + 0.5 * x, 0.5 + 0.5 * y );
                glVertex2f(x*0.5,y*0.5);
            }
            
            glEnd();
            
            return Var();
        }
        
        FHE_FUNC_IMPL(Circle,collTest)
        {
            return Var::build<bool>(arg.get<Vec2>().length() < 0.5);
        }
    }
}

